mongo reader bug: mongo cron 导致panic

## Bug Report

### logkit 相关配置：
{
"max_procs": 1,
"debug_level": 1,
"clean_self_log":true,
"bind_host":":3000",
"profile_host":":3002",
"disable_web":false,
"static_root_path":"./public",
"confs_path": ["confs*"]
}

### 运行环境: 
centos 7.2

#### logkit 版本：
最新版 20181125
### 复现过程操作步骤：

1. 设置mongo reader中的mongo cron: loop 10
2. 停止，重置任务是会报出panic

### 相关日志等其他信息：
panic: Session already closed

goroutine 112 [running]:
gopkg.in/mgo%2ev2.(*Session).cluster(0xc42013e9c0, 0x12e7688)
	/home/willjnzhang/go/src/gopkg.in/mgo.v2/session.go:1624 +0x56
gopkg.in/mgo%2ev2.(*Session).acquireSocket(0xc42013e9c0, 0xc42462f701, 0x0, 0x0, 0x0)
	/home/willjnzhang/go/src/gopkg.in/mgo.v2/session.go:4440 +0x1da
gopkg.in/mgo%2ev2.(*Query).Iter(0xc4201d7800, 0x0)
	/home/willjnzhang/go/src/gopkg.in/mgo.v2/session.go:3446 +0x1c1
github.com/qiniu/logkit/reader/mongo.(*Reader).exec(0xc42028a120, 0x12e6408, 0xc42028a120)
	/home/willjnzhang/go/src/github.com/qiniu/logkit/reader/mongo/mongo.go:379 +0x129
github.com/qiniu/logkit/reader/mongo.(*Reader).run(0xc42028a120)
	/home/willjnzhang/go/src/github.com/qiniu/logkit/reader/mongo/mongo.go:331 +0x550
github.com/qiniu/logkit/reader/mongo.(*Reader).Start.func1(0xc42028a120)
	/home/willjnzhang/go/src/github.com/qiniu/logkit/reader/mongo/mongo.go:208 +0x8d
created by github.com/qiniu/logkit/reader/mongo.(*Reader).Start
	/home/willjnzhang/go/src/github.com/qiniu/logkit/reader/mongo/mongo.go:204 +0x301

### 初步判断
runner将reader close后，因为定时任务的原因mongo reader还会异步的请求mongo，导致func exec()运行至语句iter := r.catQuery(r.collection, r.offset, r.session).Iter()时抛出panic: Session already closed。
### 初步解决方案
func exec()加recover()...... ╮(╯_╰)╭

--- a/reader/mongo/mongo.go
+++ b/reader/mongo/mongo.go
@@ -39,6 +39,11 @@ const (
 	DefaultOffsetKey = "_id"
 )
 
+const (
+	Release int32 = iota
+	NoRelease
+)
+
 type Reader struct {
 	meta *reader.Meta
 	// Note: 原子操作，用于表示 reader 整体的运行状态
@@ -50,7 +55,8 @@ type Reader struct {
 		- StatusRunning: 当前有任务正在执行
 		- StatusStopping: 数据管道已经由上层关闭，执行中的任务完成时直接退出无需再处理
 	*/
-	routineStatus int32
+	routineStatus    int32
+	isReleaseCluster int32 //Note: 原子操作，用于表示mgo session的连接池是否被释放
 
 	stopChan chan struct{}
 	readChan chan []byte //bson
@@ -72,7 +78,6 @@ type Reader struct {
 	Cron         *cron.Cron //定时任务
 	session      *mgo.Session
 	offset       interface{} //对于默认的offset_key: "_id", 是objectID作为offset，存储的表现形式是string，其他则是int64
-
 }
 
 func NewReader(meta *reader.Meta, conf conf.MapConf) (reader.Reader, error) {
@@ -110,6 +115,7 @@ func NewReader(meta *reader.Meta, conf conf.MapConf) (reader.Reader, error) {
 		meta:              meta,
 		status:            StatusInit,
 		routineStatus:     StatusInit,
+		isReleaseCluster:  Release,
 		stopChan:          make(chan struct{}),
 		readChan:          make(chan []byte),
 		errChan:           make(chan error),
@@ -283,6 +289,7 @@ func (r *Reader) Close() (err error) {
 
 	r.Cron.Stop()
 	if r.session != nil {
+		atomic.StoreInt32(&r.isReleaseCluster, Release)
 		r.session.Close()
 	}
 
@@ -364,6 +371,7 @@ func (r *Reader) exec() (err error) {
 		if err != nil {
 			return
 		}
+		atomic.StoreInt32(&r.isReleaseCluster, NoRelease)
 		r.session.SetSocketTimeout(time.Second * 5)
 		r.session.SetSyncTimeout(time.Second * 5)
 	} else {
@@ -372,11 +380,12 @@ func (r *Reader) exec() (err error) {
 			r.session.Refresh()
 			r.session.SetSocketTimeout(time.Second * 5)
 			r.session.SetSyncTimeout(time.Second * 5)
-		} else {
-			time.Sleep(time.Second * 5)
 		}
 	}
-
+	if atomic.LoadInt32(&r.isReleaseCluster) == Release {
+		log.Warnf("Runner[%s] %s mongo session's cluster has released,this exec will be exited", r.meta.RunnerName, r.Name())
+		return
+	}
 	iter := r.catQuery(r.collection, r.offset, r.session).Iter()
 
 	var result bson.M

